File: nodejs-backend/controllers/authController.js
const jwt = require('jsonwebtoken');
const { Op } = require('sequelize');
const User = require('../models/User');
const { validateRegistration, validateLogin } = require('../utils/validation');

const generateToken = (userId) => {
    return jwt.sign({ userId }, process.env.JWT_SECRET, {
        expiresIn: process.env.JWT_EXPIRES_IN || '7d'
    });
};

const register = async (req, res) => {
    try {
        const { error } = validateRegistration(req.body);
        if (error) {
            return res.status(400).json({
                success: false,
                message: error.details[0].message
            });
        }

        const { username, password } = req.body;

        const existingUser = await User.findOne({
            where: { username }
        });

        if (existingUser) {
            return res.status(409).json({
                success: false,
                message: 'User with this username already exists'
            });
        }

        const user = await User.create({
            username,
            password
        });

        const token = generateToken(user.id);

        res.status(201).json({
            success: true,
            message: 'User registered successfully',
            data: {
                user,
                token
            }
        });
    } catch (error) {
        console.error('Register error:', error);
        res.status(500).json({
            success: false,
            message: 'Internal server error during registration'
        });
    }
};

const login = async (req, res) => {
    try {
        console.log('Login request:', req.body);

        const { error } = validateLogin(req.body);
        if (error) {
            return res.status(400).json({
                success: false,
                message: error.details[0].message
            });
        }

        const { username, password } = req.body;

        const user = await User.findOne({
            where: { username: username }
        });

        console.log('Found user:', user ? `ID: ${user.id}, Username: ${user.username}` : 'No user found');

        if (!user) {
            return res.status(401).json({
                success: false,
                message: 'Invalid credentials'
            });
        }

        const isValidPassword = await user.comparePassword(password);
        console.log('Password valid:', isValidPassword);

        if (!isValidPassword) {
            return res.status(401).json({
                success: false,
                message: 'Invalid credentials'
            });
        }

        const token = generateToken(user.id);

        res.json({
            success: true,
            message: 'Login successful',
            data: {
                user,
                token
            }
        });
    } catch (error) {
        console.error('Login error:', error);
        res.status(500).json({
            success: false,
            message: 'Internal server error during login'
        });
    }
};

const getMe = async (req, res) => {
    try {
        const user = await User.findByPk(req.user.userId);

        if (!user) {
            return res.status(404).json({
                success: false,
                message: 'User not found'
            });
        }

        res.json({
            success: true,
            data: {
                user
            }
        });
    } catch (error) {
        console.error('Get me error:', error);
        res.status(500).json({
            success: false,
            message: 'Internal server error'
        });
    }
};

module.exports = {
    register,
    login,
    getMe
};

File: java-backend/src/main/java/com/example/canvasdrawingapi/controller/AuthController.java
package com.example.canvasdrawingapi.controller;

import com.example.canvasdrawingapi.dto.ApiResponse;
import com.example.canvasdrawingapi.dto.auth.AuthResponse;
import com.example.canvasdrawingapi.dto.auth.LoginRequest;
import com.example.canvasdrawingapi.dto.auth.RegisterRequest;
import com.example.canvasdrawingapi.model.User;
import com.example.canvasdrawingapi.security.JwtUtil;
import com.example.canvasdrawingapi.service.UserService;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.validation.Valid;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Optional;

@RestController
@RequestMapping("/api/auth")
public class AuthController {
    private static final Logger logger = LoggerFactory.getLogger(AuthController.class);

    @Autowired
    private UserService userService;

    @Autowired
    private JwtUtil jwtUtil;

    @PostMapping("/register")
    public ResponseEntity<ApiResponse<AuthResponse>> register(@Valid @RequestBody RegisterRequest request) {
        try {
            if (userService.existsByUsername(request.getUsername())) {
                return ResponseEntity.status(HttpStatus.CONFLICT)
                        .body(ApiResponse.error("User with this username already exists"));
            }

            User user = userService.createUser(request.getUsername(), request.getPassword());
            String token = jwtUtil.generateJwtToken(user.getId());

            AuthResponse authResponse = new AuthResponse(user, token);
            
            return ResponseEntity.status(HttpStatus.CREATED)
                    .body(ApiResponse.success("User registered successfully", authResponse));
        } catch (Exception e) {
            logger.error("Register error:", e);
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(ApiResponse.error("Internal server error during registration"));
        }
    }

    @PostMapping("/login")
    public ResponseEntity<ApiResponse<AuthResponse>> login(@Valid @RequestBody LoginRequest request) {
        try {
            logger.debug("Login request: {}", request.getUsername());

            Optional<User> userOptional = userService.findByUsername(request.getUsername());
            
            if (userOptional.isEmpty()) {
                return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                        .body(ApiResponse.error("Invalid credentials"));
            }

            User user = userOptional.get();
            logger.debug("Found user: ID: {}, Username: {}", user.getId(), user.getUsername());

            boolean isValidPassword = userService.validatePassword(user, request.getPassword());
            logger.debug("Password valid: {}", isValidPassword);

            if (!isValidPassword) {
                return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                        .body(ApiResponse.error("Invalid credentials"));
            }

            String token = jwtUtil.generateJwtToken(user.getId());
            AuthResponse authResponse = new AuthResponse(user, token);

            return ResponseEntity.ok(ApiResponse.success("Login successful", authResponse));
        } catch (Exception e) {
            logger.error("Login error:", e);
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(ApiResponse.error("Internal server error during login"));
        }
    }

    @GetMapping("/me")
    public ResponseEntity<ApiResponse<AuthResponse>> getMe(HttpServletRequest request) {
        try {
            Long userId = (Long) request.getAttribute("userId");
            
            Optional<User> userOptional = userService.findById(userId);
            
            if (userOptional.isEmpty()) {
                return ResponseEntity.status(HttpStatus.NOT_FOUND)
                        .body(ApiResponse.error("User not found"));
            }

            User user = userOptional.get();
            AuthResponse authResponse = new AuthResponse(user, null);

            return ResponseEntity.ok(ApiResponse.success("User retrieved successfully", authResponse));
        } catch (Exception e) {
            logger.error("Get me error:", e);
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(ApiResponse.error("Internal server error"));
        }
    }
}
